File: algorithms/python/streamofCharacters/streamofCharacters.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/stream-of-characters/
# Author: Miao Zhang
# Date:   2021-04-05

class TrieNode:
    def __init__(self):
        self.TrieNode = {}

    def insert(self, word):
        cur = self.TrieNode
        for w in word:
            if w not in cur:
                cur[w] = {}
            cur = cur[w]
        cur['#'] = 1

    def search(self, word):
        cur = self.TrieNode
        for w in word:
            if w not in cur:
                return False
            if '#' in cur[w]:
                return True
            cur = cur[w]
        return False


class StreamChecker:

    def __init__(self, words: List[str]):
        self.trie = TrieNode()
        self.stream = deque([])
        for word in set(words):
            self.trie.insert(word[::-1])

    def query(self, letter: str) -> bool:
        self.stream.appendleft(letter)
        return self.trie.search(self.stream)


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: algorithms/cpp/streamofCharacters/streamofCharacters.cpp
// Source: https://leetcode.com/problems/stream-of-characters/
// Author: Miao Zhang
// Date:   2021-04-05

class TrieNode {
public:
    ~TrieNode() {
        for (auto node: next_) {
            delete node;
        }
    }

    void reverseBuild(const string&s, int i) {
        if (i == -1) {
            is_word_ = true;
            return;
        }
        const int idx = s[i] - 'a';
        if (!next_[idx]) next_[idx] = new TrieNode();
        next_[idx]->reverseBuild(s, i - 1);
    }

    bool reverseSearch(const string& s, int i) {
        if (i == -1 || is_word_) return is_word_;
        int idx = s[i] - 'a';
        if (!next_[idx]) return false;
        return next_[idx]->reverseSearch(s, i - 1);
    }

private:
    bool is_word_ = false;
    TrieNode* next_[26] = {0};
};

class StreamChecker {
public:
    StreamChecker(vector<string>& words) {
        root_ = std::make_unique<TrieNode>();
        for (const string& w: words) {
            root_->reverseBuild(w, w.length() - 1);
        }
    }

    bool query(char letter) {
        s_ += letter;
        return root_->reverseSearch(s_, s_.length() - 1);
    }

private:
    string s_;
    unique_ptr<TrieNode> root_;
};

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker* obj = new StreamChecker(words);
 * bool param_1 = obj->query(letter);
 */
